src/core/ext/filters/client_channel/resolver/dns/event_engine/event_engine_client_channel_resolver.cc

```diff
@@ -0,0 +1,535 @@
+    // the resolver will never see a completed request.
+    if (hostname_handle_.has_value()) {
+      event_engine_resolver_->CancelLookup(*hostname_handle_);
+    }
+    if (srv_handle_.has_value()) {
+      event_engine_resolver_->CancelLookup(*srv_handle_);
+    }
+    for (const auto& handle : balancer_hostname_handles_) {
+      event_engine_resolver_->CancelLookup(handle);
+    }
+    if (txt_handle_.has_value()) {
+      event_engine_resolver_->CancelLookup(*txt_handle_);
+    }
+  }
+  Unref(DEBUG_LOCATION, "Orphan");
+}
+
+void EventEngineDNSRequestWrapper::OnHostnameResolved(
+    absl::StatusOr<std::vector<EventEngine::ResolvedAddress>> addresses) {
+  absl::optional<grpc_core::Resolver::Result> result;
+  {
+    MutexLock lock(&on_resolved_mu_);
+    if (orphaned_) return;
+    hostname_handle_.reset();
+    addresses_ = addresses;
+    result = OnResolvedLocked();
+  }
+  if (result.has_value()) {
+    resolver_->OnRequestComplete(std::move(*result));
+  }
+}
+
+void EventEngineDNSRequestWrapper::OnSRVResolved(
+    std::string srv_host,
+    absl::StatusOr<std::vector<EventEngine::DNSResolver::SRVRecord>>
+        srv_records) {
+  absl::optional<grpc_core::Resolver::Result> result;
+  {
+    MutexLock lock(&on_resolved_mu_);
+    if (orphaned_) return;
+    srv_handle_.reset();
+    if (srv_records.ok()) {
+      balancer_addresses_ = grpc_core::ServerAddressList();
+      // Do a subsequent hostname query if SRV records were returned
+      for (const auto& srv_record : *srv_records) {
+        balancer_hostname_handles_.insert(
+            event_engine_resolver_->LookupHostname(
+                absl::bind_front(
+                    &EventEngineDNSRequestWrapper::OnBalancerHostnamesResolved,
+                    Ref(DEBUG_LOCATION, "OnBalancerHostnamesResolved"),
+                    srv_host),
+                srv_record.host, std::to_string(srv_record.port),
+                resolver_->query_timeout_ms_));
+        GRPC_EVENT_ENGINE_DNS_TRACE(
+            "DNSResolver::%p Started resolving balancer hostname %s:%d",
+            resolver_.get(), srv_record.host.c_str(), srv_record.port);
+      }
+      return;
+    }
+    // An error has occurred, finish resolving.
+    balancer_addresses_ = srv_records.status();
+    result = OnResolvedLocked();
+  }
+  if (result.has_value()) {
+    resolver_->OnRequestComplete(std::move(*result));
+  }
+}
+
+void EventEngineDNSRequestWrapper::OnBalancerHostnamesResolved(
+    std::string authority,
+    absl::StatusOr<std::vector<EventEngine::ResolvedAddress>>
+        new_balancer_addresses) {
+  absl::optional<grpc_core::Resolver::Result> result;
+  {
+    MutexLock lock(&on_resolved_mu_);
+    if (orphaned_) return;
+    ++number_of_balancer_hostnames_resolved_;
+    if (new_balancer_addresses.ok()) {
+      if (!balancer_addresses_.ok()) {
+        GRPC_EVENT_ENGINE_DNS_TRACE(
+            "DNSResolver::%p previous balancer hostname lookup error, ignoring "
```

Comment:
This is fixed, and the errors are appended to the resolution_note. You're right, it would have crashed ... either our naming tests don't cover that, or we haven't tested them properly yet. More of that will be done with the EventEngine DNS Resolver integration that Yijie is working on.

Reply:
Pinging @yijiem for how you'd prefer to track this work. I can leave a TODO for you in the test if that'd help.

Reply:
Add a TODO sounds good. Seems like in general the naming test does not have the dataset of multiple SRV records. And for this specific case the first balancer address lookup failed while the second one succeeded?

Reply:
I added a TODO for that specific edge case. It's a bit like a regression test, given it would have been a bug if our eyes didn't catch it first.

Reply:
I think it would be pretty straightforward to fuzz this resolver, I'll look into it this morning.

Reply:
Super, thank you!